File: src/vendors/wc3maptranslator/data/Terrain.ts
interface Terrain {
  tileset: string
  customTileset: boolean
  tilePalette: string[]
  cliffTilePalette: string[]
  map: MapSize
  // "Masks"
  groundHeight: number[][]
  waterHeight: number[][]
  boundaryFlag: boolean[][]
  flags: number[][]
  groundTexture: number[][]
  groundVariation: number[][]
  cliffVariation: number[][]
  cliffTexture: number[][]
  layerHeight: number[][]
}

interface MapSize {
  width: number
  height: number
  offset: Offset
}

interface Offset {
  x: number
  y: number
}

export type { Terrain, MapSize, Offset };

export enum TerrainFlag {
  Unwalkable = 0x0002,
  Unflyable = 0x0004,
  Unbuildable = 0x0008,
  Ramp = 0x0010, // ramp flag (used to set a ramp between two layers)
  Blight = 0x0020, // blight flag (ground will look like Undead's ground)
  Water = 0x0040, // water flag (enable water)
  Boundary = 0x0080, // boundary flag 2 (used on "camera bounds" area. Usually set by the World Editor "boundary" tool.)
}

File: golang/internal/wc3/data/terrain.go
package data

// Offset is map tile offset within the full terrain grid.
type Offset struct {
	X float32
	Y float32
}

// MapSize describes terrain dimensions and offset.
type MapSize struct {
	Width  int
	Height int
	Offset Offset
}

// TerrainFlag is a war3map.w3e tile flag bit.
type TerrainFlag uint16

const (
	TerrainUnwalkable  TerrainFlag = 0x0002
	TerrainUnflyable   TerrainFlag = 0x0004
	TerrainUnbuildable TerrainFlag = 0x0008
	TerrainRamp        TerrainFlag = 0x0010
	TerrainBlight      TerrainFlag = 0x0020
	TerrainWater       TerrainFlag = 0x0040
	TerrainBoundary    TerrainFlag = 0x0080
)

// Terrain is war3map.w3e terrain data.
type Terrain struct {
	Tileset           string
	CustomTileset     bool
	TilePalette       []string
	CliffTilePalette  []string
	Map               MapSize
	GroundHeight      [][]int
	WaterHeight       [][]int
	BoundaryFlag      [][]bool
	Flags             [][]uint16
	GroundTexture     [][]int
	GroundVariation   [][]int
	CliffVariation    [][]int
	CliffTexture      [][]int
	LayerHeight       [][]int
}
